test/common_distributed.py

```diff
@@ -129,26 +127,48 @@ def setUp(self):
         super(MultiProcessTestCase, self).setUp()
         self.skip_return_code_checks = []
         self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
-        self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
+        self.file_name = tempfile.NamedTemporaryFile(delete=False).name
 
     def tearDown(self):
         super(MultiProcessTestCase, self).tearDown()
         for p in self.processes:
             p.terminate()
 
-    def _spawn_process(self, rank):
-        name = 'process ' + str(rank)
-        process = multiprocessing.Process(target=self._run, name=name, args=(rank,))
-        process.start()
-        return process
+    def _current_test_name(self):
+        # self.id() == e.g. '__main__.TestDistributed.TestAdditive.test_get_rank'
+        return self.id().split(".")[-1]
+
+    def _fork_process(self):
+        self.processes = []
+        for rank in range(int(self.world_size)):
+            process = torch.multiprocessing.Process(
+                target=self.__class__._run,
+                name='process ' + str(rank),
+                args=(rank, self._current_test_name(), self.file_name))
+            process.start()
+            self.processes.append(process)
+
+    def _spawn_process(self):
+        ctx = torch.multiprocessing.get_context('spawn')
+        self.processes = []
+        for rank in range(int(self.world_size)):
+            process = ctx.Process(
+                target=self.__class__._run,
+                name='process ' + str(rank),
+                args=(rank, self._current_test_name(), self.file_name))
+            process.start()
+            self.processes.append(process)
```

Comment:
These two do the same, but with a different context (default vs custom). Please consolidate into 1 base function.

Could you add some comment about `torch.multiprocessing.get_context` only being available on Python 3 (also see https://docs.python.org/3/library/multiprocessing.html#multiprocessing.get_context).

If I recall correctly, the Python 3 thing is why you use spawn in the RPC tests and fork in the c10d tests, right?

Reply:
1. I tried to consolidate them  by passing "ctx = get_context("fork"/"spawn")" to base function, but it failed in python2 tests, get_context() is not available in python2. 

so still separating it into two: 
for fork, call torch.multiprocessing.process(), in default it will use fork in python2 and python3

for spawn, call ctx = get_context("fork"/"spawn"), then ctx.process()

2. we can still make c10d to use spawn in python 3 test and use fork in python 2 test in following up diff

Reply:
also, the reason I did not use native multiprocessing.process() with start_method(), because it will have two weird test failures I mentioned before: 1) asan failure even I skipped asan tests 2) MKL incompatiable with numpy library.

Reply:
I was thinking something like this. It uses the `Process` class on the module for Python 2. That should work right?

```python
def _start_processes(self, proc):
  self.processes = []
  for rank in range(int(self.world_size)):
    process = proc(
      target=self.__class__._run,
      name='process ' + str(rank),
      args=(rank, self._current_test_name(), self.file_name))
    process.start()
    self.processes.append(process)

def _fork_processes(self):
  if six.PY3:
    proc = torch.multiprocessing.get_context("fork").Process
  else:
    proc = torch.multiprocessing.Process
  self._start_processes(proc)

def _spawn_processes(self):
  if six.PY3:
    proc = torch.multiprocessing.get_context("spawn").Process
  else:
    raise RuntimeError("Cannot use spawn start method with Python 2")
  self._start_processes(proc)
```

Reply:
I think those failures indicate something else is wrong. Did you end up filing issues for them?

Reply:
awesome, it works! just updated the PR, still did not get used to pass callable constructor into functions in python

Reply:
@pietern  I guess it is related to test setup on OSS? because these tests are not even ran yet, it failed at setupClass, I guess.  To file the issue, I will need to write a separate simple program and reproduce it in CI tests

Reply:
right now, I have 3 pending issues to file :), I feel it is better to write some small reproducible sample codes to file, they are still in my TODO list:

1. call sparse tensor coalesce, indices tensor is not contiguous;  
2. multiprocess test with start_method() failed in CI tests; 
3. caffe2.python._import_c_extension can not be imported in pytorch_macos_10_13_py3_test